logrotate.c

```diff
@@ -1344,49 +1342,114 @@ static long daysElapsed(const struct tm *now, const struct tm *last)
     return (long) (diff / (24 * 3600));
 }
 
+static int isSecureDirStep(const char *pathname, uid_t exempt_uid, gid_t exempt_gid)
+{
+    struct stat sb;
+
+    if (lstat(pathname, &sb) < 0)
+        return -errno;
+
+    if (S_ISLNK(sb.st_mode)) {
+        /* check whether symlink target is inside a secure directory */
+        char *canonpath;
+        int r;
+
+        canonpath = realpath(pathname, NULL);
+        if (!canonpath)
+            return -errno;
+
+        r = isSecureDir(canonpath, exempt_uid, exempt_gid);
```

Comment:
Should we limit the recursion (to depth at most 1?) so that an attacker cannot cause excessive recursion withing logrotate by changing the file system while we are traversing it?

As I understand it, there is no valid scenario where `realpath()` would be called more than once?

Reply:
There might be cases like `/var/log/myapp` -> `/var/log/myapp2.0` and `/var/log/myapp2.0/default` -> `/var/log/myapp2.0/core`.
Nevertheless added a depth limit.